pkg/metrics/metrics.go

```diff
@@ -173,13 +190,53 @@ func (m *Metrics) UpdateCertificateExpiry(crt *v1alpha1.Certificate, secretListe
 		return
 	}
 
-	updateX509Expiry(crt.Name, crt.Namespace, cert)
+	updateX509Expiry(crt, cert)
 }
 
-func updateX509Expiry(name, namespace string, cert *x509.Certificate) {
+func updateX509Expiry(crt *v1alpha1.Certificate, cert *x509.Certificate) {
 	// set certificate expiry time
 	expiryTime := cert.NotAfter
+	RegisteredCertificates.mtx.Lock()
+	defer RegisteredCertificates.mtx.Unlock()
 	CertificateExpiryTimeSeconds.With(prometheus.Labels{
-		"name":      name,
-		"namespace": namespace}).Set(float64(expiryTime.Unix()))
+		"name":      crt.Name,
+		"namespace": crt.Namespace}).Set(float64(expiryTime.Unix()))
+	RegisteredCertificates.certificates[crt] = struct{}{}
+}
+
+func (m *Metrics) CleanUp(certificateLister cmlisters.CertificateLister) {
```

Comment:
Can we un-export this method, and instead of calling it from within the Certificates controller instead have this package set up the goroutine that handles cleaning up? Whilst right now we only clean up certificate resources, it may well be the case that we have additional metric types that need cleaning up in future, and that logic should be hidden from the consumers of this package.

I guess you are going to need a reference to the `certificateLister` in the *Metrics struct, which will need to be instantiated on startup.

Reply:
Since there could be multiple instances of `Metrics` which all reference the same Prometheus structures, should the `certificateLister` be a global?

In this snippet:
https://github.com/jetstack/cert-manager/blob/3201d126d0441298805b8ff6165afebae4ce1550/cmd/controller/app/controller.go#L71
`metrics` is started but there is no `certificateLister` to pass, so I'm thinking that when the Certificates controller creates its instance of `Metrics`, it should then set the global, which will then allow clean up.
I guess the clean-up goroutine should check if the global is `nil` and do nothing before it is set.

Reply:
I added a function to set the `certificateLister`, which is called after the controller for the Certificates is created - not sure if this is the best way to set the reference?